pkg/tsdb/cloudwatch/cloudwatch.go

```diff
@@ -74,16 +76,98 @@ type cloudWatchExecutor struct {
 
 func (e *cloudWatchExecutor) newSession(region string) (*session.Session, error) {
 	dsInfo := e.getDSInfo(region)
-	creds, err := getCredentials(dsInfo)
+
+	cacheKey := fmt.Sprintf("%d:%s:%s:%s:%s:%s", dsInfo.AuthType, dsInfo.AccessKey, dsInfo.Profile, dsInfo.AssumeRoleARN,
+		dsInfo.ExternalID, region)
+
+	sessCacheLock.RLock()
+	if env, ok := sessCache[cacheKey]; ok {
+		if env.expiration.After(time.Now().UTC()) {
+			sessCacheLock.RUnlock()
+			plog.Debug("Using cached AWS session", "key", cacheKey)
+			return env.session, nil
+		}
+	}
+	sessCacheLock.RUnlock()
+
+	cfgs := []*aws.Config{
+		{
+			CredentialsChainVerboseErrors: aws.Bool(true),
+		},
+	}
+
+	var regionCfg *aws.Config
+	if dsInfo.Region == defaultRegion {
+		plog.Warn("Region is set to \"default\", which is unsupported")
+		dsInfo.Region = ""
+	}
+	if dsInfo.Region != "" {
+		regionCfg = &aws.Config{Region: aws.String(dsInfo.Region)}
+		cfgs = append(cfgs, regionCfg)
+	}
+
+	switch dsInfo.AuthType {
+	case authTypeSharedCreds:
+		plog.Debug("Authenticating towards AWS with shared credentials", "profile", dsInfo.Profile,
+			"region", dsInfo.Region)
+		cfgs = append(cfgs, &aws.Config{
+			Credentials: credentials.NewSharedCredentials("", dsInfo.Profile),
+		})
+	case authTypeKeys:
+		plog.Debug("Authenticating towards AWS with an access key pair", "region", dsInfo.Region)
+		cfgs = append(cfgs, &aws.Config{
+			Credentials: credentials.NewStaticCredentials(dsInfo.AccessKey, dsInfo.SecretKey, ""),
+		})
+	case authTypeDefault:
+		plog.Debug("Authenticating towards AWS with default SDK method", "region", dsInfo.Region)
+	default:
+		panic(fmt.Sprintf("Unrecognized authType: %d", dsInfo.AuthType))
+	}
+	sess, err := newSession(cfgs...)
 	if err != nil {
 		return nil, err
 	}
 
-	cfg := &aws.Config{
-		Region:      aws.String(dsInfo.Region),
-		Credentials: creds,
+	duration := stscreds.DefaultDuration
+	expiration := time.Now().Add(duration)
+	if dsInfo.AssumeRoleARN != "" {
+		// We should assume a role in AWS
+		plog.Debug("Trying to assume role in AWS", "arn", dsInfo.AssumeRoleARN)
+
+		cfgs := []*aws.Config{
+			{
+				CredentialsChainVerboseErrors: aws.Bool(true),
+			},
+			{
+				Credentials: newSTSCredentials(sess, dsInfo.AssumeRoleARN, func(p *stscreds.AssumeRoleProvider) {
+					// Not sure if this is necessary, overlaps with p.Duration and is undocumented
+					p.Expiry.SetExpiration(expiration, 0)
+					p.Duration = duration
+					if dsInfo.ExternalID != "" {
+						p.ExternalID = aws.String(dsInfo.ExternalID)
+					}
+				}),
+			},
+		}
+		if regionCfg != nil {
+			cfgs = append(cfgs, regionCfg)
+		}
+		sess, err = newSession(cfgs...)
+		if err != nil {
+			return nil, err
+		}
 	}
-	return newSession(cfg)
+
+	plog.Debug("Successfully authenticated towards AWS")
+
+	sessCacheLock.Lock()
+	sessCache[cacheKey] = envelope{
```

Comment:
Think it's a potential risk of `cacheKey` have been added in between `sessCacheLock.RUnlock()` and `sessCacheLock.Lock()`. So should probably have a check here to verify it's still not in cache, if it return it. WDYT?

Reply:
@marefr How do we benefit from returning any existing cache entry on line 164 instead of just overwriting it with the session that has already been created at this point? Since the session is already created, I don't see that anything is gained from not inserting it in the cache.

Reply:
Hmm. Was thinking that it means someone else/a query is currently using the session added in between RUnlock and Lock and should maybe reuse that instead? What's needed is maybe to lock earlier, before creating the new session, but could maybe create additional time for holding the lock?

Reply:
@marefr What practical negative effect does it have if a session was inserted into the cache in the meantime and it's overwritten by the new one? The only effect I can see is that the next time the cache key is looked up, a new session is returned? I might be overlooking something, but I fail to see anything bad happening?

Reply:
I've found [singleflight](https://godoc.org/golang.org/x/sync/singleflight) to be a good solution to synchronizing cache interactions. However, it is true that there is minimal downside to potentially creating two sessions and have one overwrite the other in the cache.